File: 3121. count-the-number-of-special-characters-ii.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        last_lower = {}
        first_upper = {}
        invalid = set()

        for i, ch in enumerate(word):
            letter = ch.lower()

            if ch.islower():
                last_lower[letter] = i

                # lowercase appeared after uppercase
                if letter in first_upper:
                    invalid.add(letter)

            else:
                # store only first uppercase occurrence
                if letter not in first_upper:
                    first_upper[letter] = i

        special_count = 0

        for letter in last_lower:
            if letter in first_upper and letter not in invalid:
                special_count += 1

        return special_count

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    word1 = "aaAbcBC"
    output1 = sol.numberOfSpecialChars(word1)
    print(f"Example 1: word={word1}")
    print(f"Output: {output1}")
    print(f"Expected: 3")
    print()
    
    # Example 2
    word2 = "abc"
    output2 = sol.numberOfSpecialChars(word2)
    print(f"Example 2: word={word2}")
    print(f"Output: {output2}")
    print(f"Expected: 0")
    print()
    
    # Example 3
    word3 = "AbBCab"
    output3 = sol.numberOfSpecialChars(word3)
    print(f"Example 3: word={word3}")
    print(f"Output: {output3}")
    print(f"Expected: 0")

File: 3121. count-the-number-of-special-characters-ii.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int numberOfSpecialChars(String word) {
        int[] lastLower = new int[26];
        int[] firstUpper = new int[26];

        Arrays.fill(lastLower, -1);
        Arrays.fill(firstUpper, -1);

        Set<Integer> invalid = new HashSet<>();

        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);

            if (ch >= 'a' && ch <= 'z') {
                int idx = ch - 'a';

                lastLower[idx] = i;

                if (firstUpper[idx] != -1) {
                    invalid.add(idx);
                }

            } else {
                int idx = ch - 'A';

                if (firstUpper[idx] == -1) {
                    firstUpper[idx] = i;
                }
            }
        }

        int specialCount = 0;

        for (int i = 0; i < 26; i++) {
            if (lastLower[i] != -1 &&
                firstUpper[i] != -1 &&
                !invalid.contains(i)) {

                specialCount++;
            }
        }

        return specialCount;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        String word1 = "aaAbcBC";
        int output1 = sol.numberOfSpecialChars(word1);
        System.out.println("Example 1: word=\"aaAbcBC\"");
        System.out.println("Output: " + output1);
        System.out.println("Expected: 3");
        System.out.println();
        
        // Example 2
        String word2 = "abc";
        int output2 = sol.numberOfSpecialChars(word2);
        System.out.println("Example 2: word=\"abc\"");
        System.out.println("Output: " + output2);
        System.out.println("Expected: 0");
        System.out.println();
        
        // Example 3
        String word3 = "AbBCab";
        int output3 = sol.numberOfSpecialChars(word3);
        System.out.println("Example 3: word=\"AbBCab\"");
        System.out.println("Output: " + output3);
        System.out.println("Expected: 0");
    }
}